### DISCUSSION TITLE
Passthrough output typing

### DISCUSSION THREAD

**Discussion by iambryanhaney** [colinhacks/zod#1345] [Q&A]
A zod schema can forward unknown keys to its output via `.passthrough()`, but TS won't know about them in `z.output<typeof schema>` (as expected - how would it?).

In the context of a multi-step form wizard, in order to aggregate a final output type, one could specify each schema as a merge of the previous, i.e.: 

```TSX
const stepOne = z.object({ ... });
const stepTwo = z.object({ ... }).merge(stepOne);
const stepThree = z.object({ ... }).merge(stepTwo)

type Payload = z.output<typeof stepThree>;
```

... but that would _re-run validations on known-valid fields_ at each step. To avoid this, you could do something along the lines of:

```TSX
const stepOne = z.object({ ... });
const stepTwo = z.object({ ... }).passthrough();
const stepThree = z
  .object({ ... })
  .passthrough()
  .transform((schema) => {
    const payload = stepOne.merge(stepTwo).merge(schema);
    return schema as unknown as typeof payload;
  );

type Payload = z.output<typeof stepThree>;
```

Alternatively, if you didn't want the intermediate `payload` variable:

```TSX
  .passthrough()
  .transform((schema) => schema as unknown as z.output<
      z.ZodObject<
        z.extendShape<
          z.extendShape<typeof stepOne.shape, typeof stepTwo.shape>,
          typeof stepThree.shape>
        >
     >
  >)
```

This solution feels kind of ok but also kind of bloated + antipattern to me.

It would be nice if there was something like a `z.sequence()` that enables passthrough and aggregates output:

```TSX
const wizardSchema = z.sequence([
  z.object({ firstName: z.string() }),
  z.object({ lastName: z.string() }),
  z.object({ age: z.number() }),
])

type Payload = z.output<wizardSchema>;
// { firstName: string, lastName: string, age: number }
```

I can see that as being a bit out of zod's intended scope, so I'm mostly just talking out loud and seeing if anyone has suggestions for a cleaner solution?

**Comment by jonlambert** [comment]
I know this is a late reply, but stumbled across this post while looking for a related discussion. 

> ... but that would re-run validations on known-valid fields at each step.

Doesn't this seem desirable, though? 

Presumably, you're saving the values of steps one and two somewhere, and so there's a non-zero chance of that data changing between now and then. 🙂 If it has changed, handle that logic by passing your user back to that failed step. If there are lots of async checks, cache them locally in a value-keyed store somewhere.
